### PR TITLE
Python3 internal storage for Seq, MutableSeq

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request addresses addresses a remaining issue of the Python2 -> Python3 transition.

In Python2, sequences were stored as a `str` object in the `_data` attribute of `Seq` objects. In Python 2, `str` objects store byte strings where each letter is 1 byte.

During the transition to Python3, we continued to use `str` objects. However, the closest thing in Python3 to a Python2 `str` object is not a Python3 `str` but a Python3 `bytes` object.

Using `bytes` for the `_data` attribute of `Seq` objects has three advantages:
 - Parsing sequence files to `bytes` is faster as it avoids going through a `TextIOWrapper`;
 - `bytes` supports the buffer protocol, making them much easier to use in Python and C extensions;
 - `bytes` has a direct counterpart (`bytearray`) for mutable sequences; note that our current usage of Unicode `array` objects in `MutableSeq` has been deprecated since Python 3.3.

This PR replaces `str` by `bytes` for the `_data` attribute of `Seq` objects, and `str` by `bytearray` for the `data` attribute of `MutableSeq` objects. As these are internal attributes, it should not affect any users except those that are using non-ASCII sequences in `Seq` objects. However that should not matter, as in Python2 we did not support that either.

Note that the PR does not attempt to make use of the speed and convenience improvements offered by `bytes`/`bytearray`; this is left for separate PRs.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
I thought you wanted to make ``Seq`` a subclass of ``str``? This would make that more difficult, wouldn't it?

Do you have a plan for how a binary mode parser would create a ``Seq`` object *without* turning the bytes into unicode to give to ``Seq.__init__`` only for that to turn it back again? The overhead of an ``if isinstance(...):`` might outweigh the benefits; and I would expect a try/except to be even slower. i.e. Is your first advantage real?

The ``Seq`` object's ``encode`` method can probably be simplified if we make this change (although our subclasses like ``UnknownSeq`` and ``DBSeq`` might then need their own definition).

More generally, do you think is likely to break many user defined ``Seq`` subclasses?

**Comment by peterjc** [comment]
This does seem like a chance to clean up the ``MutableSeq`` and its partial support for sequences with non-single-letter elements - see #1681

**Comment by peterjc** [comment]
How would you feel about postponing this proposal until after Biopython 1.78 is out?  That would give a good baseline for doing performance profiling against.

**Comment by mdehoon** [comment]
> I thought you wanted to make Seq a subclass of str? This would make that more difficult, wouldn't it?

With Python2, that made sense. With Python3, it doesn't, as `str` does not support the buffer protocol.

> Do you have a plan for how a binary mode parser would create a Seq object without turning the bytes into unicode to give to Seq.__init__ only for that to turn it back again? The overhead of an if isinstance(...): might outweigh the benefits; and I would expect a try/except to be even slower. i.e. Is your first advantage real?

Sorry but I don't understand your question. If `_data` is `bytes` and the parser works in binary mode, then you don't turn bytes into unicode at any point.

> More generally, do you think is likely to break many user defined Seq subclasses?

I don't think that is likely. There are few, if any, user-defined `Seq` subclasses, and in any case subclassing should not depend on the internals of how data is stored in a `Seq` object.

**Comment by mdehoon** [comment]
> This does seem like a chance to clean up the MutableSeq and its partial support for sequences with non-single-letter elements - see #1681

Yes I think `MutableSeq` should just be exactly the same as `Seq`, other than that it's mutable.

**Comment by mdehoon** [comment]
> How would you feel about postponing this proposal until after Biopython 1.78 is out? That would give a good baseline for doing performance profiling against.

That is my preferred solution, but only if Biopython 1.78 comes out very soon. I would suggest to just put a freeze on Biopython and get 1.78 out as soon as possible, such that 1.78 is 1.77 minus alphabets, and we can find out in real life if alphabet removal broke anything.

If release 1.78 cannot come out soon, then I don't want to slow down and wait. Alphabet removal was only a first step; we still have a very long way to go.

**Comment by peterjc** [comment]
One thing which frustrates me about the ``MutableSeq`` is that its reverse and (reverse)complement methods are in-place. I would be tempted to make them return a new object by default, with the option of the historic in-place behaviour. However, I'm even more tempted to deprecate and remove the class completely.

Getting back to my parser question, suppose the parser has ``sequence_in_bytes = b"ACGT"`` and then wants to call ``sequence_as_seq_object = Seq(...)`` with the current public API they have to do ``sequence_as_seq_object = Seq(sequence_in_bytes.decode("ASCII"))`` - so the bytes get turned into unicode, and then turned back again.

**Comment by mdehoon** [comment]
> Getting back to my parser question, suppose the parser has `sequence_in_bytes = b"ACGT"` and then wants to call `sequence_as_seq_object = Seq(...)` with the current public API they have to do `sequence_as_seq_object = Seq(sequence_in_bytes.decode("ASCII"))` - so the bytes get turned into unicode, and then turned back again.

The current PR is just to make sure that storing data as `bytes` does not break anything. To make the API fast, we can just let `Seq.__init__` accept `str` and `bytes`. Or you could create an empty `Seq` object and set the `_data` attribute equal to `sequence_in_bytes`. Or you could have a dedicated factory function that only takes `bytes`. It's just a matter of optimization.

**Comment by peterjc** [comment]
If you are confident this isn't going to be a net performance overhead, we should try it. It may need rewriting both parsers and writers with bytes support - particularly the FASTA and FASTQ ones where big files are now common.

But let's get Biopython 1.78 released first 👍

**Comment by mdehoon** [comment]
What is the timeline for release 1.78?

**Comment by peterjc** [comment]
August? We need the transition guidance documentation done first - the clearer that is, the less time we'll spend guiding people after the release.

Documentation review and issue triage is also worth doing first - there are likely still more issues we can now close.

File: Bio/Seq.py [hunk 1]
```diff
@@ -125,10 +124,10 @@ def __str__(self):
                 as_string = str(seq_obj)
 
         """
-        return self._data
+        return self._data.decode("ASCII")
 
     def __hash__(self):
-        """Hash of the sequence as a string (ignoring alphabet) for comparison.
```
Comment (peterjc) [inline Bio/Seq.py:131]: Can you put this in a separate PR please?

File: Bio/Seq.py [hunk 2]
```diff
@@ -1633,9 +1632,9 @@ def join(self, other):
 
 
 class MutableSeq:
-    """An editable sequence object (with an alphabet).
```
Comment (peterjc) [inline Bio/Seq.py:1636]: Can you put this in a separate PR please, along with the other removal of an alphabet mention?

**Comment by mdehoon** [comment]
OK let's aim for August. In the meantime, I will continue on this PR to move on with the next step so we don't lose time. I will convert it to draft to make clear it should not be merged yet.

**Comment by peterjc** [comment]
Sounds good.

File: Bio/Seq.py (hunk 1 above)
Comment (peterjc) [reply Bio/Seq.py:131]: Included on #3192

File: Bio/Seq.py (hunk 2 above)
Comment (peterjc) [reply Bio/Seq.py:1636]: Included on #3192

**Comment by mdehoon** [comment]
Estimated time to parse human genome chr1 in Fasta format (uncompressed):

Biopython release 1.77: 3.56 seconds
Current PR, parsing binary stream: 2.91 seconds

**Comment by mdehoon** [comment]
Note that the current PR will break any existing scripts that subclass `Seq`, `MutableSeq`, or `UnknownSeq`.

**Comment by mdehoon** [comment]
Memory usage in bytes of a Seq object of length _n_:

- Biopython ≤ 1.77 (before dropping alphabets): 328 + _n_
- Biopython 1.78 (after dropping alphabets): 271 + _n_
- Current PR, using subclassing: 177 + _n_

**Comment by peterjc** [comment]
Interesting. Running here with the master on macOS, Python 3.7.6 - the numbers are similar but not as low.

* Biopython 1.77 (and the magnitude of this surprised me): 582 + n
* Biopython master (after dropping alphabets, plan for Biopython 1.78): 287 + n
* Biopython master subclassing str (quick hack, lots of test fail): 233 + n
* Biopython subclassing bytes (this PR): 193 + n

I used ``for i in (1, 10, 100, 1000, 10000): print(get_size(Seq("A"*i)) - i)``.

It should it be easy to subclass ``str`` for comparison, but it wasn't as trivial as I thought. That approach does seem less likely to introduce subtle typing bugs, which worries me in explicitly subclassing ``bytes`` but exposing a unicode ``str`` based API. Any end user code doing an ``isinstance(...)`` check might break with this. I may be worrying about a tiny minority of use cases?

**Comment by mdehoon** [comment]
@peterjc 
Note that in Biopython 1.77, alphabets are a singleton class so the alphabet of each Seq (assuming they are using the same alphabet) is pointing to the same Alphabet object. Your numbers are correct for each Seq object, but additional Seq objects will take less memory as the alphabet is reused.

**Comment by mdehoon** [comment]
Total memory usage in bytes of a `Seq` object of length *n*:
- Without `__slots__`: 209 + *n* bytes
- With `__slots__ = ()`: 57 + *n* bytes

All tests pass in either case. Note that with `__slots__ = ()`, users cannot create a new attribute on a `Seq` object.

**Comment by mdehoon** [comment]
Proof-of-principle of `Seq` and `SeqRecord` unification, by storing `id`, `name`, `description`, `annotations`, `features`, `dbxrefs`, and `letter_annotations` directly on a `Seq` object, and removing `SeqRecord` from Biopython.

If none of these attributes are defined, memory usage of a `Seq` object of size *n* is 113 + *n* bytes with the current implementation.
This can be reduced further to 65 + *n* bytes by improving the implementation (should be straightforward but I haven't tried it).

**Comment by peterjc** [comment]
Do you see a compelling advantage to a single class, over having a minimal plain sequence class (``Seq``) and an annotated sub class (``SeqRecord``)?

**Comment by mdehoon** [comment]
@peterjc  Do you see an advantage to having two classes?

**Comment by peterjc** [comment]
Yes, I see a clear division between the two classes. One is like a basic immutable datatype (like a string), the other a mutable composite object.

The ``Seq`` is like a string with no metadata (literally so with the removal of the alphabet), making it very general and suitable for use in 3rd party data structures. The logic for sorting and comparison is also absolutely clear (behaves like a string). It behaves like a Python immutable, and is hashable for use as a dictionary key etc.

 The ``SeqRecord`` supports rich annotation, but the design is somewhat GenBank/EMBL centric, especially the feature implementation. It is not immutable, cannot be hashed, and does not attempt to define equality and sorting.

In your combined class, sequence based equality seems like a recipe for confusion - would ``"ACGT"`` and ``Seq("ACGT")`` and ``Seq("ACGT", name="Example)`` and ``Seq("ACGT", letter_annotations={"phred_quality": [40,40,39,40]})`` really be equal?

**Comment by mdehoon** [comment]
> The Seq is like a string with no metadata (literally so with the removal of the alphabet)

Exactly. Then what is the point of having a Seq class?

I find our current scheme with two classes confusing. If you have ever spent an afternoon trying to figure out why your code didn't work, just to find out that you were using a SeqRecord and not a Seq object, you'd understand why.

Btw our documentation in places confuses Seq and SeqRecord, not to mention that SeqIO parses SeqRecords, not Seq objects.

**Comment by mdehoon** [comment]
> Do you see a compelling advantage to a single class, over having a minimal plain sequence class (Seq) and an annotated sub class (SeqRecord)?

> In your combined class, sequence based equality seems like a recipe for confusion - would "ACGT" and Seq("ACGT") and Seq("ACGT", name="Example) and Seq("ACGT", letter_annotations={"phred_quality": [40,40,39,40]}) really be equal?

But how would this work if SeqRecord subclasses Seq?

**Comment by peterjc** [comment]
If working with DNA or RNA, the ``Seq`` object has a lot of nice biological methods. It is less compelling for proteins though.

If you've found places in the documentation mixing up ``Seq`` and ``SeqRecord``, could you log them?

In hindsight, naming ``Bio.SeqIO`` matching BioPerl was a mistake. Changing module names to be lower case pep8 compliant is a big change which could be part of a Biopython v2.

If ``SeqRecord`` was to subclass ``Seq`` (which as I recall you suggested in the past), then I would suggest it override ``__eq__`` and ``__hash__`` etc to be not implemented.

**Comment by mdehoon** [comment]
> If SeqRecord was to subclass Seq (which as I recall you suggested in the past), 

Yes I think I did at some point. But after having thought more about it and trying some things, I see endless problems with this approach.

> then I would suggest it override __eq__ and __hash__ etc to be not implemented.

Then I I have two SeqRecords, and I want to compare their sequence contents, I would have to copy them first.

**Comment by peterjc** [comment]
If you want to compare the sequence of two ``SeqRecord`` objects, currently it would be ``record1.seq == record2.seq``, maybe that would have to remain the same even if subclassing?

Perhaps SeqRecord subclassing Seq is not going to be practical.

**Comment by mdehoon** [comment]
> If you want to compare the sequence of two SeqRecord objects, currently it would be record1.seq == record2.seq, maybe that would have to remain the same even if subclassing?

Exactly. Then `record.seq` is a new object with a copy of the sequence contained in `SeqRecord`.

**Comment by mdehoon** [comment]
> In your combined class, sequence based equality seems like a recipe for confusion - would "ACGT" and Seq("ACGT") and Seq("ACGT", name="Example) and Seq("ACGT", letter_annotations={"phred_quality": [40,40,39,40]}) really be equal?

Actually that is consistent with Python's behavior in general:
```
>>> class Seq(str):
...     pass
... 
>>> 
>>> a = Seq("ABCD")
>>> b = Seq("ABCD")
>>> a == b
True
>>> a.name = "seq1"
>>> b.name = "seq2"
>>> a == b
True
>>> 
```
We could add a specific method (`equals`, maybe) to test equality of all attributes, not just the sequence contents.

**Comment by peterjc** [comment]
I accept that if we have to define ``SeqRecord`` comparison, doing it based on the sequence alone is the least bad option (other than the default of object identity of being the same instance, which is what it does now, and is not very useful).

It will cause confusion though, e.g. comparing FASTQ derived records which have the same sequence but different qualities. Or a FASTA and GenBank derived record with the same sequence most of their annotation being different.

(Biopython had a similar discussion long ago when we changed the ``Seq`` comparison to ignore the alphabet; this change seems more dramatic though)

**Comment by mdehoon** [comment]
It depends a bit on our choice of class names and variable names. If a unified class is called `Seq`, and the examples in the documentation using `sequence` as the variable name, then it may not be so surprising that `sequence1 == sequence2` compares the sequence contents only.

**Comment by mdehoon** [comment]
Btw in Biopython 1.78, the memory used by a `SeqRecord` with a sequence of length *n*, with no attributes, is 1759 + *n*.

**Comment by mdehoon** [comment]
@peterjc  Was I able to convince you of the need to unify `Seq` and `SeqRecord`?

Btw the `vars` built-in function can be used to check whether the attributes of two objects are equal.
For example,
```
>>> class Seq(str): pass
... 
>>> a = Seq("ABCD")
>>> b = Seq("ABCD")
>>> a == b
True
>>> a.name = 'seq1'
>>> b.name = 'seq2'
>>> a == b
True
>>> vars(a)
{'name': 'seq1'}
>>> vars(b)
{'name': 'seq2'}
>>> vars(a) == vars(b)
False
```

**Comment by peterjc** [comment]
The ``vars`` built-in function would be useful. That would probably address the other potential equality checking people might want if SeqRecord subclassed Seq and did pure sequence comparison.

What about mutable vs immutable and hashing of a SeqRecord though (if subclassed or otherwise combined into a single class)?

**Comment by mdehoon** [comment]
> What about mutable vs immutable and hashing of a SeqRecord though (if subclassed or otherwise combined into a single class)?

The hash is based on the sequence content only, and since the sequence content doesn't change, the class is immutable. This again follows the usual practice in Python:
```python
>>> class Seq(str): pass
... 
>>> a = Seq("ABCD")
>>> b = Seq("ABCD")
>>> d = {}
>>> d[a] = 'hello'
>>> d[b]
'hello'
>>> a.name = 'seq1'
>>> b.name = 'seq2'
>>> d[a]
'hello'
>>> d[b]
'hello'
>>> len(d)
1
```

**Comment by peterjc** [comment]
Hmm. Maybe. It feels wrong though.

I'd rather block adding attributes to ``Seq`` (admittedly a behaviour change) as part of making it more truly immutable and more like ``str`` (and potentially have ``Seq`` subclass ``str``), with ``SeqRecord`` remaining separate.

**Comment by Lucioric2000** [comment]
Hello. I know that much code uses x.seq == y syntax for SeqRecord, but if you unify the Seq and SeqRecord class you can solve that by using a seq descriptor to give the widely used .seq attibute (making this descriptor to raise a deprecation warning). Best regards.

Enviado desde mi Samsung Mobile de Telcel
Obtener Outlook para Android<https://aka.ms/ghei36>

________________________________

**Comment by mdehoon** [comment]
@peterjc 
But the current approach is not acceptable either. Too cumbersome, too confusing, at times infuriating. How can it be improved, if not by unifying `Seq` and `SeqRecord`?

**Comment by peterjc** [comment]
It seems your concerns with the ``SeqRecord`` are very different from mine. I like the current separation of a sequence class, and and a separate class for an annotated sequence.

As noted above, I would consider making ``Seq`` truly immutable (block adding attributes at run time), and perhaps subclass string. And I would consider removing ``MutableSeq``.

For the ``SeqRecord``, it is the details of the annotation storage that I might seek to change in a backward compatibility breaking change (e.g. having ``.name`` and ``.id`` is generally redundant, I would pick just one, and for each format link this to what ever is easiest for indexing the file - generally coming from the first line of the record; ``.dbxrefs`` should allow for linked data with URLs; ``.features`` needs to be indexed, filterable, potentially hierarchical - not just a flat list which is a poor fit for GFF).

**Comment by mdehoon** [comment]
@peterjc  Yes I know that you like the current separation into two classes, but I have not yet heard a good justification of that. Currently we have three ways to represent a sequence in Biopython: A plain string, a `Seq` object, and a `SeqRecord` object. This means implementing sequence methods (e.g.`reverse_complement` and `translate`) three times (not counting `MutableSeq` and `UnknownSeq`). This already indicates that our design is suboptimal, to say the least. It also means that functions acting on sequences have to be prepared to accept all three variants, and for users to remember which function expects what.
This is one example of how quickly things go pear-shaped with our current design:
```python
>>> from Bio import SeqIO
>>> sequence1 = SeqIO.read("1MRR_A.gp", 'genbank')
>>> sequence2 = SeqIO.read("protein_refseq.gb", 'genbank')
>>> from Bio.pairwise2 import align
>>> alignments = align.globalxx(sequence1, sequence2)
>>> alignments[0].score
462.0
>>> 
```

**Comment by peterjc** [comment]
Making ``Seq`` subclass ``str`` would help here. Also we try to encourage using the methods over the functions.

Your example is surprising - although the ``SeqIO`` documentation does deliberately use ``record`` as a variable name:

```pycon
>>> from Bio import SeqIO
>>> record1 = SeqIO.read("1MRR_A.gp", 'genbank')
>>> record2 = SeqIO.read("protein_refseq.gb", 'genbank')
>>> from Bio.pairwise2 import align
>>> alignments = align.globalxx(record1.seq, record2.seq)
>>> alignments[0].score
90.0
>>> alignments = align.globalxx(record1, record2) #  bad
>>> alignments[0].score
462.0
```

I see this is aligning the ``__str__`` output of the ``SeqRecord``, which is bad. It probably deserves a bug report of its own.

**Comment by peterjc** [comment]
As an aside, I think type annotation would help here - it is not just for documentation.

**Comment by mdehoon** [comment]
> Making Seq subclass str would help here.

That would break `UnknownSeq`.

**Comment by mdehoon** [comment]
Here is another example:
```
>>> from Bio.SeqRecord import SeqRecord
>>> record = SeqRecord("ACGTCA", id='seq1')
>>> record.translate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/biopython-1.78-py3.6-macosx-10.9-x86_64.egg/Bio/SeqRecord.py", line 1322, in translate
    table=table, stop_symbol=stop_symbol, to_stop=to_stop, cds=cds, gap=gap
TypeError: translate() takes no keyword arguments
>>> 
```

**Comment by peterjc** [comment]
Was that a reference to #2968, making the ``Seq`` translate method compatible with the string method of the same name?

**Comment by mdehoon** [comment]
> (And why do you think that example is a problem? - you get a nice clear error)

Except that the error message is wrong. The bug is where the `SeqRecord` was created with a plain string; it should have been created with a `Seq` object:
```python
>>> from Bio.SeqRecord import SeqRecord
>>> from Bio.Seq import Seq
>>> record = SeqRecord(Seq("ACGTCA"), id='seq1')
>>> record.translate()
SeqRecord(seq=Seq('TS'), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])
```

**Comment by peterjc** [comment]
Point taken. I'd suggest adding another type check to the SeqRecord ``__init__`` there.

**Comment by mdehoon** [comment]
@peterjc  Wouldn't be easier if `SeqRecord` accepted a plain string?

**Comment by peterjc** [comment]
If you'd be happy with the ``SeqRecord`` automatically upgrading a plain string to a ``Seq``, then sure.

**Comment by mdehoon** [comment]
But why do that if a string can do the job?

**Comment by peterjc** [comment]
A string does not have the biological methods which are assumed by the ``SeqRecord`` biological method implementations (although that could be worked around), and by many downstream code using the ``SeqRecord``.

**Comment by mdehoon** [comment]
But the `SeqRecord` itself does, so that doesn't matter.

**Comment by mdehoon** [comment]
This example shows how `UnknownSeq` will break if `Seq` is a subclass of string:
```python
>>> class Seq(str): pass
... 
>>> class UnknownSeq(Seq): 
...     def __new__(cls, length):
...         self = super().__new__(cls, "")
...         self.length = length
...         return self
...     def __str__(self):
...         return "N" * self.length
... 
>>> u = UnknownSeq(6)
>>> print(u)
NNNNNN
>>> str(u)
'NNNNNN'
>>> "ABCD" + u
'ABCD'
>>> 
```
The same bug will occur for subclasses of `Seq` that implement a lazy-loading scheme for large (genome-sized) sequence files.

**Comment by peterjc** [comment]
The could be other implications for the ``UnknownSeq`` if ``Seq`` subclassed string, but it does already defines its own ``__add__`` etc so that should work: https://github.com/biopython/biopython/blob/biopython-178/Bio/Seq.py#L1161

**Comment by mdehoon** [comment]
I was just using addition as an example here. Without special treatment of `UnknownSeq`s, any function that acts on a string will interpret the `UnknownSeq` as an empty string. Here is a different example:

```python
>>> class Seq(str): pass
... 
>>> class UnknownSeq(Seq):
...     def __new__(cls, length):
...         self = super().__new__(cls, "")
...         self.length = length
...         return self
...     def __str__(self):
...         return "N" * self.length
... 
>>> u = UnknownSeq(6)
>>> print(u)
NNNNNN
>>> str(u)
'NNNNNN'
>>> u in "ABCD"
True
>>> 
```

**Comment by peterjc** [comment]
Yes, but again we've implemented ``__contains__`` already on the ``UnknownSeq``. I think all of this was necessitated by subclassing ``Seq``, which is so string like already that I struggle to see what new problems subclassing string would introduce.

**Comment by mdehoon** [comment]
`__contains__` won't save you here. It gets invoked only for the opposite case (`"ABCD" in u`):

```
>>> class Seq(str): 
...     def __contains__(self, char):
...         print("In __contains__")
...         return str(char) in str(self)
... 
>>> class UnknownSeq(Seq):
...     def __new__(cls, length):
...         self = super().__new__(cls, "") 
...         self.length = length
...         return self
...     def __str__(self):
...         return "N" * self.length
... 
>>> u = UnknownSeq(6)
>>> print(u)
NNNNNN
>>> str(u)
'NNNNNN'
>>> "ABCD" in u
In __contains__
False
>>> "NNNNNN" in u
In __contains__
True
>>> u in "ABCD"  # __contains__ won't get called here
True
>>>
```

**Comment by mdehoon** [comment]
This is another example (continuing from the previous one):
```python
>>> print(u)
NNNNNN
>>> "ABCDNNNNNN".index(u)
0
```

**Comment by mdehoon** [comment]
or this one...
```python
>>> print(u)
NNNNNN
>>> "ABCD".join("NNNNNN")
'NABCDNABCDNABCDNABCDNABCDN'
>>> "ABCD".join(u)
''
```

**Comment by peterjc** [comment]
Ah. Is this because in that proof of principle the ``UnknownSeq`` gets initialised as an empty sequence ``""``? I don't see an easy way round that - a dummy value could have other side effects. Perhaps ``UnknownSeq`` would no longer be able to subclass ``Seq`` if that subclassed ``str``. Might want to use an ABC here for adding the biological methods?

**Comment by mdehoon** [comment]
Whatever you use to initialize the `UnknownSeq` will be used as *the* string value of the `UnknownSeq`. There is no way around that. 

> Might want to use an ABC here for adding the biological methods?

 I think you first need to define which problem you are trying to fix by subclassing `Seq` from string.

**Comment by peterjc** [comment]
It seemed like making ``Seq`` subclass ``str`` would be an easy way to have all the string methods available without explicitly having to implement them all. Perhaps ``collections.UserString`` might work better? https://docs.python.org/3/library/collections.html#collections.UserString

**Comment by mdehoon** [comment]
But why would you need all those methods? And you would have to override most of those methods anyway because the `str` won't quite do what you'd want for `Seq`. And you would be adding a bunch of methods to `Seq` that don't make sense for biological sequences (like `capitalize`).

**Comment by mdehoon** [comment]
Using
```python
import resource
from Bio import SeqIO

handle = open("SRR12143416.fasta")
records = SeqIO.parse(handle, 'fasta')
records = list(records)
handle.close()

memory = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

print(memory)
```
to evaluate the memory usage needed to store 1,000,000 RNA-seq sequences from ENCODE.

Biopython 1.78: 1,564,688,384 bytes
After unifying `Seq` and `SeqRecord`: 488,103,936 bytes, so a reduction by 69%.

Same information is being stored in both cases.

**Comment by peterjc** [comment]
Impressive. The big saving are presumably not setting up empty lists and dictionaries for the annotation. Could the same trick be applied to the ``SeqRecord`` directly (even if it did require C code)?

Can you do the list of two strings equivalent as a lower limit?

```python
with open("SRR12143416.fasta") as handle:
    records = list(SimpleFastaParser(handle))
```

**Comment by mdehoon** [comment]
> Can you do the list of two strings equivalent as a lower limit?
```python
with open("SRR12143416.fasta") as handle:
    records = list(SimpleFastaParser(handle))
```
This gives 417,857,536 bytes.
Then with Biopython 1.78, 73% of memory usage is overhead.
After unifying `Seq` and `SeqRecord`, 14% of memory usage is overhead.
